tailnet/tailnettest/tailnettest.go

```diff
@@ -61,3 +63,60 @@ func RunDERPAndSTUN(t *testing.T) *tailcfg.DERPMap {
 		},
 	}
 }
+
+// RunDERPOnlyWebSockets creates a DERP mapping for tests that
+// only allows WebSockets through it. Many proxies do not support
+// upgrading DERP, so this is a good fallback.
+func RunDERPOnlyWebSockets(t *testing.T) *tailcfg.DERPMap {
+	logf := tailnet.Logger(slogtest.Make(t, nil))
+	d := derp.NewServer(key.NewNode(), logf)
+	handler := derphttp.Handler(d)
+	var closeFunc func()
+	handler, closeFunc = tailnet.AddWebsocketSupport(d, handler)
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/derp" {
+			handler.ServeHTTP(w, r)
+			return
+		}
+		if r.Header.Get("Upgrade") != "websocket" {
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write([]byte(fmt.Sprintf(`Invalid "Upgrade" header: %s`, html.EscapeString(r.Header.Get("Upgrade")))))
+			return
+		}
+		handler.ServeHTTP(w, r)
+	}))
+	server.Config.ErrorLog = tslogger.StdLogger(logf)
+	server.Config.TLSNextProto = make(map[string]func(*http.Server, *tls.Conn, http.Handler))
+	server.StartTLS()
+	t.Cleanup(func() {
+		server.CloseClientConnections()
+		server.Close()
+		d.Close()
+		closeFunc()
```

Comment:
```suggestion
		closeFunc()
		d.Close()
```

A more natural order? (Similar to defering.)